README.md
# Coverex

Coverex is an Elixir Coverage tool used by `mix`. It provides tables with overviews of
module and function coverage data, includings links to annotated source code files.

[![Build Status](https://travis-ci.org/alfert/coverex.svg?branch=master)](https://travis-ci.org/alfert/coverex)
[![Coverage Status](https://coveralls.io/repos/alfert/coverex/badge.png?branch=master)](https://coveralls.io/r/alfert/coverex?branch=master)
[![hex.pm version](https://img.shields.io/hexpm/v/coverex.svg?style=flat)](https://hex.pm/packages/coverex)

## Configuration

Coverex is completely configured via `mix.exs` of your project. To enable `Coverex`,
you add this line to your `mix.exs` file

	test_coverage: [tool: Coverex.Task]

as part of the regular project settings. In addition to that, you need to add Coverex
to the dependencies of your project. Coverex is available via `hex.pm`, so you need only to
add this line to the dependencies in your `mix.exs` file:

	{:coverex, "~> 1.4.10", only: :test}

For debugging purposes, the log level can be set as addition to the `tool` option. The default
value is `:error`. To set the log level to `:debug` you use this line in your `mix.exs` file:

	test_coverage: [tool: Coverex.Task, log: :debug]

The usual log levels of the `Logger` application of Elixir are available.

If you set the flag `coveralls: true` and you are running on `travis-ci`, the coverage information are sent to http://coveralls.io . An example configuration would be

	test_coverage: [tool: Coverex.Task, coveralls: true]


Since `coverex 1.4.7`, a summary on the module level is printed on the console. You can switch this off by setting the option `console_log` to `false`.

	test_coverage: [tool: Coverex.Task, console_log: false]

If you want to ignore some of the modules to appear in the coverage reports, e.g
because they are generated automatically and therefore often needs no test coverage,
you can do this since `coverex 1.4.8`. Use the option `ignore_modules` and assign
to it a list of module names to ignore.

	test_coverage: [tool: Coverex.Task, ignore_modules: [Database, Database.User]]

Since `coverex 1.4.10` the from the Elixir compiler automatically generated
functions `__info__` and `__struct__` are removed from the list of covered
functions. It simply makes no sense to include them into the list (see also #24).

## Running Coverex

If you have configured Coverex as described above you can run Coverex as a drop-in replacement
for the regular coverage mechanism of mix:

    $> mix test --cover

The coverage reports are found in the `cover` directory or what you have configured as coverage directory
as explained in the docs of the `Mix.Tasks.Test` task.

## Contributing

Please use the GitHub issue tracker for

* bug reports and for
* submitting pull requests

## License

Coverex is provided under the Apache 2.0 License.


CHANGELOG.md
# Coverex Changelog

## 1.5.0
* Moving up to Elixir 1.6 and later
* [Support Poison 4.0](https://github.com/alfert/coverex/pull/47)

## 1.4.15
* Fixes the broken 1.4.14 package (https://github.com/alfert/coverex/issues/42), found by https://github.com/aseigo

## 1.4.14
* [Skip non-Elixir modules](https://github.com/alfert/coverex/pull/41), thanks to https://github.com/reachfh
* [Fix inconsistent indentation](https://github.com/alfert/coverex/pull/49) thanks to https://github.com/hauleth

## 1.4.13
* [update dependency to accept Poison version 3.0](https://github.com/alfert/coverex/pull/39),
  thanks to https://github.com/carlo-colombo
* Little changes to the release process

## 1.4.12
* [Added totals in output](https://github.com/alfert/coverex/pull/36), thanks to https://github.com/henb

## 1.4.11
* [Coverage numbers were being reported on essentially random lines.](https://github.com/alfert/coverex/pull/35), thanks to https://github.com/scouten


test/coverex_test.exs
defmodule CoverexTest do
	use ExUnit.Case

	alias Coverex.Task

	test "coveralls wished?" do
	  	opts = [coveralls: true]
	  	assert Task.post_to_coveralls?(opts)
	end

	test "coveralls not wished?" do
	  	opts = [coveralls: false]
	  	refute Task.post_to_coveralls?(opts)
	end

	test "is coveralls requested on the commandline?" do
		conf = Mix.Project.config()
		IO.inspect conf[:test_coverage]
		assert Task.post_to_coveralls?(conf[:test_coverage]),
			"coveralls must be set on the commandline to make a positive test"
	end

	test "positive check environment for travis" do
		env = %{"TRAVIS" => "true"}
		assert Task.running_travis?(env)
	end

	test "negative check environment for travis" do
		env = %{"TRAVIS" => "no"}
		refute Task.running_travis?(env)

		env = %{"TRAVIS" => "TRUE"}
		refute Task.running_travis?(env)
	end

	# test "post to whatever" do
	# 	conf = Mix.Project.config()
	# 	assert nil != conf[:test_coverage], "requires to run under `test --cover` "

	# 	assert :ok = Task.post_coveralls([Coverex.Task], ".", "id-123", "http://pastebin.com")
	# end
end


test/coverex_source_test.exs
defmodule CoverexSourceTest do

	use ExUnit.Case
  require Logger

	@mod Coverex.Source

  	test "compile info " do
  		info = Coverex.Source.get_compile_info(@mod)

      assert is_list(info)
      path = Keyword.get(info, :source) |> to_string()
      assert Path.extname(path) == ".ex"
      assert Path.basename(path, ".ex") == "source"
  	end

  	test "source info" do
  		source = Coverex.Source.get_source_path(@mod)

  		assert is_list(source)
  	end

  	test "get source and quoted" do
  		{quoted, source} = Coverex.Source.get_quoted_source(@mod)

  		assert is_binary(source)
  		assert {:defmodule, [line: 1], _} = quoted
  	end

  	test "check 1 alias" do
  		as = Coverex.Source.alias_mod(X)
  		assert [:X] = as
      m = Coverex.Source.alias_to_atom(as ++ [:Elixir])
      assert X == m
  	end

  	test "check 2 aliases" do
  		as = Coverex.Source.alias_mod(@mod)
  		assert [:Coverex, :Source] == as
      m = Coverex.Source.alias_to_atom([:Elixir | as] |> Enum.reverse)
      assert @mod == m
  	end

  	test "find a single module" do
  		{:ok, quoted} = generate_mod("X", ["f", "g", "hs"]) |> Code.string_to_quoted
  		# IO.puts "quoted code is \n #{inspect quoted}"

      mods = Coverex.Source.find_all_mods_and_funs(quoted)
      # IO.inspect mods
      assert mods[X][X] == 1
  	end

    test "find all modules and funs" do
      ms = %{"X" => X, "Y" => Y, "A.B.C" => A.B.C}
      funs = [{X, :f, 0}, {X, :g, 0}, {X, :hs, 0}]
      {:ok, mod} = generate_mod(Map.keys(ms), funs) |> Code.string_to_quoted
      Logger.debug("mod = #{inspect mod}")
      all_mods = Coverex.Source.find_all_mods_and_funs(mod)
      Logger.debug("all_mods = #{inspect all_mods}")

      ms |> Map.values |> Enum.each(fn(mod_name) ->
        assert %{} = all_mods[mod_name]
        assert is_integer(all_mods[mod_name][mod_name])
        funs |> # Enum.map(&String.to_atom/1) |>
          Enum.each(fn({_m, f, a}) ->
            # IO.puts "f = #{inspect f}"
            assert is_integer(all_mods[mod_name][{mod_name, f, a}])
          end)
      end)
    end

    test "generate line entries" do
      m = [{X, 1},
        {{X, :f, 2}, 5},
        {{X, :g, 0}, 9}] |>
          Enum.into(%{})
      cover = [{{X, 5}, 1}, {{X, 6}, 1}, {{X, 7}, 1}, {{X, 9}, 3}]
      lines = Coverex.Source.generate_lines(cover, m)

      f_link = Coverex.Task.module_anchor({X, :f, 2})
      # IO.puts "Lines = #{inspect lines}"
      assert %{} = lines
      assert {1, f_link} == lines[5]
      assert {1, nil} = lines[6]
    end

    test "find Protocols and implementations" do
      src = """
      defprotocol Disposable do
        def dispose(disposable)
      end
      defimpl Disposable, for: Function do
        def dispose(fun), do: fun.()
      end
      """

      {:ok, mods} = src |> Code.string_to_quoted
      # Logger.debug("mods: #{inspect mods}")
      all_mods = Coverex.Source.find_all_mods_and_funs(mods)
      # Logger.debug("all mods: #{inspect all_mods}")

      assert %{} = all_mods[Disposable]
      assert %{} = all_mods[Disposable.Function]
      assert is_integer(all_mods[Disposable][Disposable])
      assert is_integer(all_mods[Disposable.Function][Disposable.Function])
    end

    test "nested Modules" do
      src = """
      defmodule X.Y do
        defmodule Z do
          def g(a), do: 1 + a
        end
      end
      """
      {:ok, mods} = src |> Code.string_to_quoted
      Logger.debug("mods: #{inspect mods}")
      all_mods = Coverex.Source.find_all_mods_and_funs(mods)
      Logger.debug("all mods: #{inspect all_mods}")

      refute :erlang.is_map(all_mods[X])
      assert %{} = all_mods[X.Y]
      assert %{} = all_mods[X.Y.Z]
      assert is_integer(all_mods[X.Y.Z][{X.Y.Z, :g, 1}])
    end

    test "nested modules with structs and types" do
      src = """
      defmodule X.Y do

        @type signal :: Signal.t(any) # {:Signal, reference, pid, any}
        @type signal(a) :: Signal.t(a) # {:Signal, reference, pid, a}
        # @type sig_func(a, b) :: ((a, any) -> ({:reply, b, any} | {:noreply, any})) when a: var, b: var
        @type sig_func :: ((any, any) -> ({:reply, any, any} | {:noreply, any}))

        defmodule Signal do
          @derive Access
          defstruct id: nil, # id of the invividual signal
            source: nil, # pid of the source
            value: nil # current value of the signal

          @type t(a) :: %__MODULE__{id: nil|reference, source: nil|pid, value: nil | a}
        end
        def f(x), do: x+1
      end
      """
      {:ok, mods} = src |> Code.string_to_quoted
      Logger.debug("mods: #{inspect mods}")
      all_mods = Coverex.Source.find_all_mods_and_funs(mods)
      Logger.debug("all mods: #{inspect all_mods}")

      assert %{} = all_mods[X.Y]
      assert %{} = all_mods[X.Y.Signal]
      assert %{} = all_mods[Access.X.Y.Signal]
      assert is_integer(all_mods[X.Y][{X.Y, :f, 1}])
    end

    test "Merge coverage data of a single file" do
      mod = [a: [{2, nil}, {3, 1}, {4, 5}, {5, 2}]]
      merged = Coverex.Source.merge_coverage(mod)
      assert %{1=> nil, 2=> nil, 3=> 1, 4=> 5, 5=> 2} == merged
    end

    test "Merge coverage data of a set of files" do
      mod = [a: [{2, nil}, {3, 1}, {4, 5}, {5, 2}],
             b: [{8, 4}, {9, 2}, {10, 1}]
          ]
      merged = Coverex.Source.merge_coverage(mod)
      assert %{1 => nil, 2=> nil, 3=> 1, 4=> 5, 5=> 2,
               6 => nil, 7=>nil,
               8 => 4, 9 => 2, 10 => 1} == merged
    end

  	@doc """
  	Generates modules with functions as binaries
  	"""
  	def generate_mod(mod), do: generate_mod(mod, [])
  	def generate_mod(mod, funs) when is_binary(mod) do
  		fs = funs |> Enum.reduce("", fn(f, acc) -> make_fun(f) <> acc end)
  		"""
  		defmodule #{mod} do
		""" <> fs <>
		"""
  		end
  		"""
  	end
  	def generate_mod(mods, funs) when is_list(mods) do
  		mods |> Enum.reduce("", fn(m, acc) -> generate_mod(m, funs) <> acc end)
  	end

    defp make_fun({_, f, 0}), do: "def #{f}(), do: :funny0\n"
    defp make_fun({_, f, args}), do: "def #{f}(#{1..args |> Enum.map_join(",", &("args#{&1}"))}), do: :funny#{args}\n"
    defp make_fun(f), do: "def #{f}(), do: :funny_0\n"

end


release.sh
#!/bin/bash
# release.sh old new
set -x

# old is always with -dev
old="1.5.0"
new="1.5.1-dev"
# do not set any variables beyond this line
old_version="$old-dev"
release_version="$old"
new_version="$new"
tag_name=v$release_version
set +x

git branch | grep '* master' > /dev/null
if [ 1 -eq $? ]; then
	echo "ERROR: Not on branch master"
	exit 1
fi

read -p "Check the variables. Press Ctrl-C for exit, return for continuing"

ORIG='.orig'

# update version in mix.exs
sed -i $ORIG "s/\(version: \"\)$old_version\",/\\1$release_version\",/" mix.exs

# add to git
git commit -m "bump version to $release_version" mix.exs

# tag the commit
git tag -a -m "new release version v$release_version" v$release_version
echo "updated mix file is committed and tagged with v$release_version"

# Upload to Hex.PM (both package and docs)
echo "Publish now to Hex"
mix hex.publish

# update version in mix.exs
sed -i $ORIG "s/\(version: \"\)$release_version\",/\\1$new_version\",/" mix.exs

# add to git
git commit -m "bump version to $new_version" mix.exs
echo "updated mix file is committed with version $new_version"

# push to github
echo "Now pushing all changes to GitHub"
git push origin master --tags


lib/coverex/task.ex
defmodule Coverex.Task do
  require EEx

  require Logger

  @type mod_cover :: [{String.t, {integer, integer}}]
  @type fun_cover :: [{mfa, {integer, integer}}]

  @doc """
  Starts the `Coverex` coverage data generation. An additional option
  is

      log: :error

  which sets the log-level of the Elixir Logger application. Default value
  is `:error`. For debugging purposes, it can be set to :debug. In this case,
  the output is quite noisy...
  """
  def start(compile_path, opts) do
    Mix.shell.info "Coverex compiling modules ... "
    Mix.shell.info "compile_path is #{inspect compile_path}"
    Mix.shell.info "Options are: #{inspect opts}"
    :cover.start
    :cover.compile_beam_directory(compile_path |> to_charlist)

    if :application.get_env(:cover, :started) != {:ok, true} do
      :application.set_env(:cover, :started, true)
    end

    output = opts[:output]
    fn() ->
      Mix.shell.info "\nGenerating cover results ... "
      Application.ensure_started(:logger)
      Logger.configure(level: Keyword.get(opts, :log, :error))
      File.mkdir_p!(output)

      # determine all modules to analyze
      all_modules = :cover.modules
                    |> filter_modules(Keyword.get(opts, :ignore_modules, []))

      Enum.each all_modules, fn(mod) ->
        # :cover.analyse_to_file(mod, '#{output}/#{mod}.1.html', [:html])
        write_html_file(mod, output)
      end
      {mods, funcs} = coverage_data(all_modules)
      {mods_sum, total} = calculate_summary(mods)

      write_module_overview(mods_sum, total, output)
      write_function_overview(funcs, output)
      generate_assets(output)
      # Text overview output
      if Keyword.get(opts, :console_log, true) do
        puts_module_overview(mods_sum, total)
      end
      # ask for coveralls option
      if (running_travis?() and post_to_coveralls?(opts)) do
        post_coveralls(:cover.modules, output, travis_job_id())
      end
    end
  end

  def filter_modules(all_modules, ignore_list) do
    all_modules
    |> Enum.reject(fn m ->
      ignore_list |> Enum.member?(m)
    end)
    |> Enum.filter(fn m ->
      case Atom.to_string(m) do
        "Elixir." <> _rest -> true
        _ -> false
      end
    end)
  end

  @doc "is post to coveralls requested?"
  def post_to_coveralls?(nil), do: post_to_coveralls?([])
  def post_to_coveralls?(opts) do
    Keyword.get(opts, :coveralls, false)
  end

  @spec running_travis?() :: boolean
  @spec running_travis?(%{optional(String.t) => String.t}) :: boolean
  def running_travis?(env \\ System.get_env()) do
    case env["TRAVIS"] do
      "true" -> true
      _ -> false
    end
  end

  @doc "gets the travis job id out of the environment"
  @spec travis_job_id(%{String.t => String.t}) :: String.t
  def travis_job_id(env \\ System.get_env()) do
    env["TRAVIS_JOB_ID"]
  end

  @spec post_coveralls([atom], String.t, String.t, String.t) :: :ok
  def post_coveralls(mods, output, job_id, url \\ "https://coveralls.io/api/v1/jobs") do
    IO.puts "post to coveralls"
    Application.ensure_all_started(:hackney)
    source = Coverex.Source.coveralls_data(mods)
    body = Poison.encode!(%{
      :service_name => "travis-ci",
      :service_job_id => job_id,
      :source_files => source
    })
    filename = "./#{output}/coveralls.json"
    File.write(filename, body)
    response = send_http(url, filename, body)
    IO.puts("Response: #{inspect response}")
  end

  def send_http(url, filename, _body) do
    :hackney.post(url, [],
                  {:multipart, [
                    {:file, filename,
                      {"form-data", [{"name", "json_file"}, {"filename", filename}]},
                      [{"Content-Type", "application/json"}]
                    }
                  ]},
                  [:with_body])
  end

  def write_html_file(mod, output) do
    {entries, source} = Coverex.Source.analyze_to_html(mod)
    {:ok, s} = StringIO.open(source)
    lines = Stream.zip(numbers(), IO.stream(s, :line)) |> Stream.map(fn({n, line}) ->
      case Map.get(entries, n, nil) do
        {count, anchor} -> {n, {encode_html(line), count, anchor}}
        nil -> {n, {encode_html(line), nil, nil}}
      end
    end) |> Enum.map(&(&1))

    # IO.inspect(lines)
    content = source_template(mod, lines)
    File.write("#{output}/#{mod}.html", content)
  end

  # all positive numbers
  defp numbers(), do: Stream.iterate(1, &(&1+1))

  def encode_html(s, acc \\ "")
  def encode_html("", acc), do: acc
  def encode_html(s, acc) do
    case String.next_grapheme(s) do
      {">", rest} -> encode_html(rest, acc <> "&gt;")
      {"<", rest} -> encode_html(rest, acc <> "&lt;")
      {"&", rest} -> encode_html(rest, acc <> "&amp;")
      nil -> acc
      {first, rest} -> encode_html(rest, acc <> first)
    end
  end

  def write_module_overview(modules, total, output) do
    mods = Enum.map(modules, fn({mod, v}) -> {module_link(mod), v} end)
    content = overview_template("Modules", mods, total)
    File.write("#{output}/modules.html", content)
  end

  def puts_module_overview(modules_summary, total) do
    # mods = Enum.map(modules, fn({mod, v}) -> {module_link(mod), v} end)
    content = overview_text_template(modules_summary, total)

    IO.puts content
  end

  def write_function_overview(functions, output) do
    {fun_sum, total} = calculate_summary(functions)

    funs = Enum.map(fun_sum, fn({{m,f,a}, v}) -> {module_link(m, f, a), v} end)
    content = overview_template("Functions", funs, total)
    File.write("#{output}/functions.html", content)
  end

  defp module_link(mod), do: "<a href=\"#{mod}.html\">#{mod}</a>"
  defp module_link(m, f, a), do: "<a href=\"#{m}.html##{m}.#{f}.#{a}\">#{m}.#{f}/#{a}</a>"

  def module_anchor({m, f, a}), do: "<a name=\"##{m}.#{f}.#{a}\"></a>"
  def module_anchor(m), do: "<a name=\"##{m}\"></a>"

  def cover_class(nil), do: "irrelevant"
  def cover_class(0), do: "not_covered"
  def cover_class(_n), do: "covered"

  @doc """
  Returns detailed coverage data `{mod, mf}` for all modules from the `:cover` application.
  The special functions `__info__` and `__struct__` are filtered out of
  the list of covered functions.

  ## The `mod` data
  The `mod` data is a list of pairs: `{modulename, {no of covered lines, no of uncovered lines}}`

  ## The `mf` data
  The `mf` data is list of pairs: `{{m, f, a}, {no of covered lines, no of uncovered lines}}`

  """
  @spec coverage_data([atom]) :: {mod_cover, fun_cover}
  def coverage_data(all_modules) do
    modules = Enum.map(all_modules, fn(mod) ->
      {:ok, {m, {cov, noncov}}} = :cover.analyse(mod, :coverage, :module)
      {m, {cov, noncov}}
    end) |> Enum.sort
    mfunc = Enum.flat_map(all_modules, fn(mod) ->
      {:ok, funcs} = :cover.analyse(mod, :coverage, :function)
      funcs
    end)
    |> Enum.reject(fn {{_m, f, _}, _}
    when f in [:__info__, :__struct__,
               :__behaviour_info__, :__protocol__]-> true
      _ -> false end)
      |> Enum.sort
    {modules, mfunc}
  end

  ## Generate templating functions via EEx, borrowd from ex_doc
  templates = [
    overview_text_template: [:entries, :total],
    overview_entry_text_template: [:entry, :cov, :uncov, :ratio],
    overview_template: [:title, :entries, :total],
    overview_entry_template: [:entry, :cov, :uncov, :ratio],
    source_template: [:title, :lines],
    source_line_template: [:number, :count, :source, :anchor]
  ]
  Enum.each templates, fn({ name, args }) ->
    filename = Path.expand("templates/#{name}.eex", __DIR__)
    EEx.function_from_file :def, name, filename, args
  end

  # generates asset files
  defp generate_assets(output) do
    Enum.each assets(), fn({ pattern, dir }) ->
      output = "#{output}/#{dir}"
      File.mkdir output

      Enum.map Path.wildcard(pattern), fn(file) ->
        base = Path.basename(file)
        File.copy file, "#{output}/#{base}"
      end
    end
  end

  # assets are javascript, css and gif resources
  defp assets do
    [{templates_path("css/*.css"), "css"},
     {templates_path("js/*.js"), "js"},
     {templates_path("css/*.gif"), "css"}]
  end

  defp templates_path(other), do: Path.expand("templates/#{other}", __DIR__)

  # returns the average percentage, the amount of covered, the amount of uncovered
  defp calculate_summary(modules) do
    {m, {t_p, t_c, t_u}} = Enum.map_reduce(modules, {0, 0, 0}, fn({e, {c, u}}, acc) ->
      {t_percent, t_covered, t_uncovered} = acc
      p = (c / (c + u) * 100) |>  Float.round(1)

      {{e, {p, c, u}}, {t_percent + p, t_covered + c, t_uncovered + u}}
    end)

    if t_p > 0 do
      t_p = t_p
            |> (&//2).(modules |> Enum.count)
            |> Float.round(1)

      {m, {t_p, t_c, t_u}}
    else
      {m, {t_p, t_c, t_u}}
    end
  end
end


lib/coverex/source.ex
defmodule Coverex.Source do
  @moduledoc """
  This module provides access to the source code the system to be analyzed.
  """

  require Logger

  @type symbol :: :atom
  @type filename :: String.t
  @type line_pairs :: %{symbol => pos_integer}
  @type modules :: %{symbol => line_pairs}
  @type line_entries :: %{pos_integer => {pos_integer | nil, binary | nil}}
  @type source_file :: %{name: String.t, source: String.t, coverage: [pos_integer | nil]}
  @type lines :: {pos_integer, pos_integer | nil}

  @spec analyze_to_html(symbol) :: {line_entries, binary}
  def analyze_to_html(mod) when is_atom(mod) do
    Logger.debug("analyze_to_html of module #{inspect mod}")
    {quoted, source} = get_quoted_source(mod)
    mods = find_all_mods_and_funs(quoted)
    if mod == Observable.PID, do:
      Logger.debug("Mods and funs found: #{inspect mods}")
    {:ok, cover} = :cover.analyse(mod, :calls, :line)
    ## cover is [{{mod, line}, count}]
    # cover |> Enum.each &Logger.info/1
    {generate_lines(cover, mods[mod]), source}
  end

  @doc """
  Returns the coverall data for the list of mods as Elixir datastructure.
  This can be encoded as JSON for uploading to coveralls.
  """
  @spec coveralls_data([symbol]) :: [source_file]
  def coveralls_data(mods) do
    mc = mods |> Enum.map(fn(mod) -> {mod, cover_per_mod(mod)} end)
    sources_and_lines(mc) |>
      Enum.reduce([], fn({path, cover}, acc) ->
        [%{:name => filter_cwd_prefix(path),
          :source => File.read!(path),
          :coverage => cover |> lines_to_list} | acc]
      end)
  end

  @doc "Strips the current directory from the path"
  def filter_cwd_prefix(path) do
    Path.relative_to_cwd(path)
  end

  @doc """
  This function aggregates the coverage information per module to a coverage
  information per source file.

  Takes a list of modules and determines the list of corresponding filenames.
  Returns to each filename a map of coverage information for the entire file.
  """
  @spec sources_and_lines([{symbol, [lines]}]) :: [{filename, line_entries}]
  def sources_and_lines(mods) do
    # identify all modules of a source file
    # mod_files is %{path => [symbol]}
    mod_files = mods |>
      Enum.map(fn({mod, lines}) -> {mod, get_source_path(mod), lines} end) |>
      Enum.reduce(%{}, fn({m, p, l}, acc) ->
        Map.update(acc, p, [{m, l}], fn(old) -> [{m, l} | old] end)
      end)
    # for each source file, grab all coverage information on line basis,
    # merge them for all modules and fill up any leaks with nils
    mod_files |> Enum.map(fn({path, mods}) -> {path, merge_coverage(mods)} end)
  end

  @doc """
  Gets a list of all modules within one sourcefile. Calculates the coverage
  data for each module and merges them together. Returns a mapping of line number
  to coverage data for the entire source file. Guarantees that all line numbers up
  to the maximun reached line are filled in.
  """
  @spec merge_coverage([{symbol, lines}]) :: line_entries
  def merge_coverage(mods) do
    unmerged = mods |> Enum.map(fn({_mod, lines}) -> lines end) |> List.flatten
    # unmerged is [{line, count}]
    merged = unmerged |> Enum.reduce(%{}, fn ({line, count}, acc) ->
      Map.update(acc, line, count,
        fn(nil) -> count
           (c)  -> count + c end)
      end)
    # fill all gaps with nil
    max_lines = merged |> Map.keys |> Enum.max
    1..max_lines |> Enum.reduce(merged, fn(index, acc) ->
      Map.put_new(acc, index, nil) end)
  end

  @spec lines_to_list(line_entries) :: [pos_integer | nil]
  def lines_to_list(lines) do
    lines
    |> Enum.sort(fn({l1, _}, {l2, _}) -> l1 <= l2 end)
    |> Enum.drop(1) # drop the mythical line 0
    |> Enum.map(fn({_l, count}) -> count end)
  end


  @spec cover_per_mod(symbol) :: [lines]
  def cover_per_mod(mod) do
    ## cover is [{{mod, line}, count}]
    {:ok, cover} = :cover.analyse(mod, :calls, :line)
    cover |> Enum.map(fn({{_m, line}, count}) -> {line, count} end)
  end


  @spec generate_lines([{{symbol, pos_integer}, pos_integer}], line_pairs) :: line_entries
  def generate_lines(cover, nil) do
    # This seems to be a situation where Macros are used extensively.
    # Does no harm but is annoying.
    Logger.debug "mod_entry is nil and cover = #{inspect cover}"
    %{}
  end
  def generate_lines(cover, mod_entry) do
    lines_anchors = mod_entry
    |> Enum.map(fn({sym, line_nr}) ->
      {line_nr, {nil, Coverex.Task.module_anchor(sym)}} end)
    |> Enum.into(%{})

    _lines_covered = cover
    |> Enum.map(fn({{_mod, line_nr}, count}) ->
      {line_nr, {count, nil}} end)
    |> Enum.into(%{})
    |> Map.merge(lines_anchors, fn _k, {c, _}, {_, a} -> {c, a} end)
  end

  @doc """
  Returns all modules and functions together with their start lines as they definend
  in the given quoted code
  """
  @spec find_all_mods_and_funs(any) :: modules
  def find_all_mods_and_funs(qs) do
    # Logger.info qs
    acc = %{:Elixir => %{}}
    do_all_mods([:Elixir], qs, acc)
  end

  # Walks the syntax tree
  # First parameter is a reverse list of nested module scopes, i.e.
  #    X.Y is encoded as [:Y, :X, :Elixir]
  defp do_all_mods(mod_reversed, {:defmodule, [line: ln], [{:__aliases__, _, mod_name} | body]}, acc) do
    # Logger.debug ("+++ Found module #{inspect mod_name}")
    mod_alias = Enum.reverse(mod_name) ++ mod_reversed
    mod = alias_to_atom(mod_alias)
    do_all_mods(mod_alias, body, acc |> Map.put(mod, %{} |> Map.put(mod,ln)))
  end
  defp do_all_mods(mod_reversed, {:defprotocol, [line: ln], [{:__aliases__, _, mod_name} | body]}, acc) do
    Logger.debug "### found defprotocol #{inspect mod_name}"
    mod_alias = Enum.reverse(mod_name) ++ mod_reversed
    mod = alias_to_atom(mod_alias)
    do_all_mods(mod_alias, body, acc |> Map.put(mod, %{} |> Map.put(mod,ln)))
  end
  defp do_all_mods(_mod_reversed, {:defimpl, [line: ln], [{:__aliases__, _, impl_name},
      [for: {:__aliases__, _, mod_name}] | body]}, acc) do
    Logger.debug "### found defimpl #{inspect impl_name} - #{inspect mod_name}"
    # impls are always toplevel modules?!
    mod_alias = Enum.reverse(mod_name) ++ Enum.reverse(impl_name) ++ [:Elixir] # mod_reversed
    mod = alias_to_atom(mod_alias)
    do_all_mods(mod_alias, body, acc |> Map.put(mod, %{} |> Map.put(mod,ln)))
  end
  defp do_all_mods(mod_reversed, {:def, [line: ln], [{fun_name, _, nil}, body]}, acc) do
    # Logger.debug ("--- Found function #{inspect fun_name}")
    m = alias_to_atom(mod_reversed)
    do_all_mods(m, body, acc |> put_in([m, {m, fun_name, 0}], ln))
  end
  defp do_all_mods(mod_reversed, {:def, [line: ln], [{fun_name, _, args}, body]}, acc) do
    # Logger.debug ("--- Found function #{inspect fun_name}")
    m = alias_to_atom(mod_reversed)
    do_all_mods(m, body, acc |> put_in([m, {m, fun_name, length(args)}], ln))
  end
  defp do_all_mods(mod_reversed, {:@, [line: ln], [{:derive, _, [{:__aliases__, _, [:Access]}]}]}, acc) do
    # derives a new module, where Access is put in front of the module name:
    # X.Y.Z @derive Access ==> Access.X.Y.Z <<-->> [:Z,:Y,:X,:Access, :Elixir]
    # This requires that we insert :Access before :Elixir in mod_reversed
    m = ((mod_reversed |> Enum.take(length(mod_reversed) - 1)) ++ [:Access, :Elixir]) |> alias_to_atom
    acc |> Map.put(m, %{} |> Map.put(m, ln))
  end
  defp do_all_mods(m, {:__block__, _, tree}, acc) when is_list(tree), do: do_all_mods(m, tree, acc)
  defp do_all_mods(m, {:do, tree}, acc), do: do_all_mods(m, tree, acc)
  defp do_all_mods(_m, t = {_t1, _t2, _t3}, acc) do
    Logger.debug "#### Found triple #{inspect t}"
    acc
  end
  defp do_all_mods(_m, [], acc), do: acc
  defp do_all_mods(m, [ head | tree], acc) do
    # basic recursion of the tree
    acc1 = do_all_mods(m, head, acc)
    do_all_mods(m, tree, acc1)
  end
  defp do_all_mods(_m, t, acc) do
    Logger.debug ">>> Found tree #{inspect t}"
    acc
  end


  @doc "Returns the aliased module name if there are any dots in its name"
  def alias_mod(mod) when is_atom(mod) do
    mod |> Atom.to_string|> String.split(".") |>
      Enum.drop(1) |> # first element contains "Elixir" which is not needed here!
      Enum.map(&String.to_atom/1)
  end

  @doc "Returns the atom module name based on the (reversed) alias list"
  def alias_to_atom(a) when is_list(a) do
    a |> Enum.reverse |> Enum.map_join(".", &Atom.to_string/1) |> String.to_atom
  end


  @doc "Returns the quoted code and the source of a module"
  @spec get_quoted_source(atom) :: {Macro.t, binary}
  def get_quoted_source(mod) do
    path = get_source_path(mod)
    {:ok, source} = File.read(path)
    {:ok, quoted} = Code.string_to_quoted(source)
    {quoted, source}
  end


  @spec get_source_path(atom) :: binary
  def get_source_path(mod) when is_atom(mod) do
    get_compile_info(mod) |> Keyword.get(:source)
  end

  @spec get_compile_info(atom) :: [{atom, term}]
  def get_compile_info(mod) when is_atom(mod) do
    {^mod, beam, _filename} = :code.get_object_code(mod)
    case :beam_lib.chunks(beam, [:compile_info]) do
      {:ok, {^mod, [{:compile_info, compile}]}} -> compile
      _ -> []
    end
  end

end


mix.exs
defmodule Coverex.Mixfile do
  use Mix.Project

  def project do
    [app: :coverex,
     version: "1.5.1-dev",
     elixir: ">= 1.6.0",
     package: package(),
     name: "Coverex - Coverage Reports for Elixir",
     source_url: "https://github.com/alfert/coverex",
     homepage_url: "https://github.com/alfert/coverex",
     docs: [readme: "README.md", extra: "CHANGELOG.md"],
     description: description(),
     test_coverage: [tool: Coverex.Task, coveralls: true, ignore_modules: []],
     deps: deps()]
  end

  # Configuration for the OTP application
  #
  # Type `mix help compile.app` for more information
  def application do
    [ extra_applications: [:logger, :hackney]]
  end

  # List all dependencies in the format:
  #
  # {:foobar, git: "https://github.com/elixir-lang/foobar.git", tag: "0.1"}
  #
  # Type `mix help deps` for more examples and options
  defp deps do
    [
      {:hackney, "~> 1.5"},
      {:poison, "~> 3.0 or ~> 3.1 or ~> 4.0"},
      {:earmark, "~> 1.0", only: :dev},
      {:ex_doc, "~> 0.13", only: :dev},
      {:dialyxir, "~> 1.0.0-rc3", only: [:dev, :test], runtime: false},
    ]
  end

  # Hex Package description
  defp description do
    """
    Coverex is an Elixir Coverage tool used by mix. It provides tables with overviews of
    module and function coverage data, includings links to annotated source code files and
    supports coveralls.io.
    """
  end

  # Hex Package definition
  defp package do
    [maintainers: ["Klaus Alfert"],
     licenses: ["Apache 2.0"],
     links: %{"GitHub" => "https://github.com/alfert/coverex"}
    ]
  end
end
